Draft (#1)
newaperio/bamboo-viability: README.md

> # Bamboo I Viability Assessment

Reviewer: Feel like you might want to deliver this as a Google Doc on our NA letter head.

Author: @lleger Will do once I've got all of the comments here resolved.

- [Overview](#overview)
- [Problems areas](#problem-areas)
  - [Data access](#data-access)
  - [Design components](#design-components)
  - [Out-of-band decisions](#out-of-band-decisions)
  - [Participant accounts](#participant-accounts)
- [Tasks for CGA](#tasks-for-cga)
- [Architecture diagram](#architecture-diagram)
- [Prototype](#prototype)

## Overview

NewAperio has completed a viability assessment for the Bamboo I project (hereafter referred to as _Bamboo_) to ensure that the necessary technology requirements can be met. After a deep dive into the existing [Common Ground for Action (CGA)][cga] application and data model, and collaboration with the [Every Voice Engaged (EVE)][eve] technology team, we’ve uncovered a well-defined problem space and have come up with solutions that will allow the project to move forward.

> ## Problem areas

Reviewer: ```suggestion
## Problem Areas
```

https://github.com/newaperio/guides/tree/main/style/markdown#headings

These are the problem areas that we identified, and are described in further detail later in the document:

1. Data access
2. Design components
3. Out-of-band decisions
4. Participant accounts

> ### Data access

Reviewer: ```suggestion
### Data Access
```

> #### The problem

Reviewer: ```suggestion
#### Problem
```

> Data essential to the Bamboo experience is tied into the CGA environment and not exposed via a web API or other such mechanism. This data includes things like:

Reviewer: ```suggestion
Data essential to the Bamboo experience is tied into the CGA environment and not exposed via a web API or other externally-accessible mechanism. This data includes things like:
```

- Issue guides
- Options
- Actions
- Drawbacks

> #### The solution

Reviewer: ```suggestion
#### Solution
```

> NewAperio will build a proxy server application that will connect directly to CGA’s database instance via a read-only connection and expose the required data over a well-defined GraphQL web API.

Reviewer: I would add this, but this will require EVE's infra to add a Read-Replica to their stack.

Also potentially worth mentioning somewhere that NA will deliver a container that can be deployed into their K8s cluster (so little to no config/devops work necessary on their part).

We discussed other options with the CGA team that would involve them making non-neglible changes to the exising legacy codebase:

- Experimenting with [Lift’s REST helpers][lift-rest]
- Building a bespoke [JSON:API] service

> After evaluating the tradeoffs, both teams decided that the solution we chose would be best--it ensures we have access to the data Bamboo needs and requires no modification of the legacy app.

Reviewer: ```suggestion
After evaluating the tradeoffs, both teams decided that the solution we chose would be best — it ensures we have access to the data Bamboo needs and requires no modification of the legacy app.
```

https://github.com/newaperio/guides/tree/main/style/prose#prose

Reviewer: Though maybe a colon is better:

```suggestion
After evaluating the tradeoffs, both teams decided that the solution we chose would be optimal: it ensures we have access to the data Bamboo needs and requires no modification of the legacy app.
```

> ### Design components

Reviewer: ```suggestion
### Design Components
```

> #### The problem

Reviewer: ```suggestion
#### Problem
```

Some of the content included in the [prototype designs] are not available in the CGA application. This includes:

- Issue guide cover images
- Issue guide categories

> #### The solution

Reviewer: ```suggestion
#### Solution
```

> After consulting with our senior designer, we feel like we can build Bamboo without this content and not unduly compromise the user experience.

Reviewer 2: ```suggestion
After consulting with our senior designer, we feel like we can build Bamboo without this content, but it will have some impact on user experience. Without issue guide thumbnail images, the main screen of the app will be text-heavy. Without categories, quickly skimming to understand what topic an issue guide covers may be more challenging.
```

Thanks for asking! I think it _does_ have some impact on user experience and am reframing this way to make it clearer to whoever reads what the impact will be (namely that it's harder to skim/will feel more like a wall of text you have to read, rather than an easy screen to browse quickly).

Author: Thanks, @glynnis. This is perfect.

This content, though not essential, would be nice to have and we’ve identified a couple of potential pathways for getting access to it:

- The CGA team could add fields to the legacy app to allow issue guide authors to specify an image and category

> - The [EVE website][eve], a WordPress app, has the content we need and there are plugins for WordPress that can provide a drop-in web API. The difficult part of this approach, however, is linking up the WordPress data with the correct CGA issue guides.

Reviewer: I think WP has custom fields, they could add a custom field with the database ID?

Author: @lleger Yep, that would be a good way to handle it.

> As both of these options require work from the CGA team, and not much work on our part, we feel that ...

Reviewer 2: @ngscheurich Not sure if you meant to trail off here or if this is a TODO

Author: Whoops! Thanks.

> ### Out-of-band decisions

Reviewer: ```suggestion
### Out-of-band Decisions
```

> #### The problem

Reviewer: ```suggestion
#### Problem
```

In the current CGA app, participant decisions, e.g. chosen options, actions, evaluations, are recorded as part of an event stream during the forum experience. Bamboo records these decisions outside of the forum experience and CGA has not yet decided how they will be reconciled with the the event stream paradigm.

> ### The solution

Reviewer: ```suggestion
### Solution
```

> NewAperio will store user decisions as part of the Bamboo application. When a complete set of decisions has been recorded, Bamboo will send a webhook event to CGA with information about how to retrieve the records. This decouples Bamboo’s implementation from the CGA app, leaving the CGA team to design their solution without restrictions.

Reviewer: Since it's inside the same VPC in AWS, we could also use something like SQS so the CGA app can just read it at its leisure instead of us forcing it to act on an HTTP request. Whichever they prefer, not sure it matters to us.

> ### Participant accounts

Reviewer: ```suggestion
### Participant Accounts
```

> #### The problem

Reviewer: ```suggestion
#### Problem
```

> Persistent user accounts are not part of the existing CGA application, that is to say, a new user is generated per participant, per forum and does not exist out of that context. CGA is working on implementing peristent accounts but there is no planned solution for ingesting external account data.

Reviewer: ```suggestion
Persistent user accounts are not part of the existing CGA application; that is to say, a new user is generated per participant, per forum and does not exist outside of that context. CGA is working on implementing peristent accounts but there is no planned solution for ingesting external account data.
```

> #### The solution

Reviewer: ```suggestion
#### Solution
```

> Bamboo will implement its own user authentication system, which will allow users to sign up for multiple forums without having to log in each time. When webhook events are sent to CGA, indentifying user information will be included. Like the decisions themselves, this somewhat decouple the two apps and allows CGA to build a solution without inteference or restriction.

Reviewer 2: ```suggestion
Bamboo will implement its own user authentication system, which will allow users to sign up for multiple forums without having to log in each time. When webhook events are sent to CGA, identifying user information will be included. Like the decisions themselves, this will somewhat decouple the two apps and will allow CGA to build a solution without interference or restriction.
```

> Bamboo will implement its own user authentication system, which will allow users to sign up for multiple forums without having to log in each time. When webhook events are sent to CGA, indentifying user information will be included. Like the decisions themselves, this somewhat decouple the two apps and allows CGA to build a solution without inteference or restriction.

Reviewer: ```suggestion
Bamboo will implement its own user authentication system, which will allow users to sign up for multiple forums without having to log in each time. When webhook events are sent to CGA, indentifying user information will be included. Like the decisions themselves, this somewhat decouples the two apps and allows CGA to build a solution without inteference or restriction.
```

> ## Architecture diagram

Reviewer: ```suggestion
## Architecture Diagram
```

Below is a diagram describing key parts of the Bamboo architecture:

- Server app
- Web front end
- Read-only database connection
- GraphQL API
- Webhook events

> <p align="center"><img src="architecture.png" width="80%"></p>

Reviewer 2: Maybe this was intentional, but wasn't expecting to see HTML in this otherwise-Markdown doc.

Author: GitHub-flavored Markdown is cool with HTML snippets!
